### PR TITLE
[FIRRTL][InferWidths] Make solveExpr Iterative

### PR DESCRIPTION
**Pull request by prithayan**
The `solveExpr` implementation of `InferWidths` was recursive and causing stack overflow issues on some designs. 
This PR implements an iterative version of the algorithm.

Co-authored-by: Schuyler Eldridge <schuyler.eldridge@sifive.com>

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/InferWidths.cpp [hunk 1]
```diff
@@ -822,95 +822,139 @@ computeBinary(ExprSolution lhs, ExprSolution rhs,
...
+          if (solvedExprs.contains(expr->lhs()))
+            return setSolution(computeBinary(
+                solvedExprs[expr->lhs()], solvedExprs[expr->rhs()],
+                [](int32_t lhs, int32_t rhs) { return std::max(lhs, rhs); }));
+
+          worklist.push_back({expr->lhs(), indent + 1});
+          worklist.push_back({expr->rhs(), indent + 1});
+        })
+        .Case<MinExpr>([&](auto *expr) {
+          if (solvedExprs.contains(expr->lhs()))
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/InferWidths.cpp:944]: Is it necessary to check `expr->rhs()` as well? I think it's necessary to check if expressions are shared across multiple expressions.

File: lib/Dialect/FIRRTL/Transforms/InferWidths.cpp [hunk 2]
```diff
@@ -822,95 +822,142 @@ computeBinary(ExprSolution lhs, ExprSolution rhs,
...
+                              unsigned defaultWorklistSize) {
+
+  struct Frame {
+    Expr *expr;
+    unsigned indent;
+  };
+
+  // indent only used for debug logs.
+  unsigned indent = 1;
+  std::vector<Frame> worklist({{expr, indent}});
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/InferWidths.cpp:834]: nit: Is there a specific reason to use std::vector over SmallVector?

Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/InferWidths.cpp:834]: I don't think so, the worklist can grow pretty large, we can switch to smallvector, how does the logic for `reserve` look to you ? I reserved the number of exprs/2 size for this.

**Comment by prithayan** [comment]
The pre-merge-ci tests were clean, going to merge this, for the next circt release.

File: lib/Dialect/FIRRTL/Transforms/InferWidths.cpp (hunk 2 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/InferWidths.cpp:834]: I don't have good intuition for the constraint but exprs/2 sounds reasonable.

**Pull request merged**